python/pyspark/ml/clustering.py

```diff
@@ -20,9 +20,150 @@
 from pyspark.ml.wrapper import JavaEstimator, JavaModel
 from pyspark.ml.param.shared import *
 from pyspark.mllib.common import inherit_doc
+from pyspark.mllib.stat.distribution import MultivariateGaussian
 
 __all__ = ['BisectingKMeans', 'BisectingKMeansModel',
-           'KMeans', 'KMeansModel']
+           'KMeans', 'KMeansModel',
+           'GaussianMixture', 'GaussianMixtureModel']
+
+
+class GaussianMixtureModel(JavaModel, JavaMLWritable, JavaMLReadable):
+    """
+    .. note:: Experimental
+
+    Model fitted by GaussianMixture.
+
+    .. versionadded:: 2.0.0
+    """
+
+    @property
+    @since("2.0.0")
+    def weights(self):
+        """
+        Weights for each Gaussian distribution in the mixture, where weights[i] is
+        the weight for Gaussian i, and weights.sum == 1.
+        """
+        return self._call_java("weights")
+
+    @property
+    @since("2.0.0")
+    def gaussians(self):
+        """
+        Array of MultivariateGaussian where gaussians[i] represents
+        the Multivariate Gaussian (Normal) Distribution for Gaussian i.
+        """
+        return [
+            MultivariateGaussian(gaussian[0], gaussian[1])
+            for gaussian in self._call_java("gaussiansPyDump")]
+
+
+@inherit_doc
+class GaussianMixture(JavaEstimator, HasFeaturesCol, HasPredictionCol, HasMaxIter, HasTol, HasSeed,
+                      HasProbabilityCol, JavaMLWritable, JavaMLReadable):
+    """
+    .. note:: Experimental
+
+    GaussianMixture clustering.
+
+    >>> from pyspark.mllib.linalg import Vectors
+
+    >>> data1 = [(Vectors.dense([-0.1, -0.05 ]),),
+    ...          (Vectors.dense([-0.01, -0.1]),),
+    ...          (Vectors.dense([0.9, 0.8]),),
+    ...          (Vectors.dense([0.75, 0.935]),),
+    ...          (Vectors.dense([-0.83, -0.68]),),
+    ...          (Vectors.dense([-0.91, -0.76]),)]
+    >>> df1 = sqlContext.createDataFrame(data1, ["features"])
+    >>> gaussianmixture = GaussianMixture(k=3, tol=0.0001,
+    ...                                    maxIter=50, seed=10)
+    >>> model = gaussianmixture.fit(df1)
+    >>> weights = model.weights
+    >>> len(weights)
+    3
+    >>> gaussians = model.gaussians
+    >>> len(gaussians)
+    3
+    >>> transformed = model.transform(df1).select("features", "prediction")
+    >>> rows = transformed.collect()
+    >>> rows[0].prediction == rows[2].prediction
+    False
+    >>> rows[4].prediction == rows[5].prediction
+    True
+    >>> rows[1].prediction == rows[5].prediction
+    False
+    >>> rows[2].prediction == rows[3].prediction
+    True
+    >>> gmm_path = temp_path + "/gmm"
+    >>> gaussianmixture.save(gmm_path)
+    >>> gaussianmixture2 = GaussianMixture.load(gmm_path)
+    >>> gaussianmixture2.getK()
+    3
+    >>> model_path = temp_path + "/gmm_model"
+    >>> model.save(model_path)
+    >>> model2 = GaussianMixtureModel.load(model_path)
+    >>> model2.weights == model.weights
+    True
+    >>> model2.gaussians == model.gaussians
```

Comment:
Can you please move this comparison of gaussian values to a unit test in ml/tests.py?  It's too complex to include in the documentation.